### PR TITLE
Feature/perodic canister

### PR DESCRIPTION
**Pull request by Robka99**

### CODE REVIEW

File: canisters/src/vts/src/lib.rs [hunk 1]
```diff
@@ -110,6 +115,15 @@ pub struct Telemetry {
     pub t_type: TelemetryType,
 }
 
+#[derive(BEncode, BDecode, Debug, PartialEq, Eq, CandidType, Deserialize, Clone, Default)]
```
Comment (lavrd) [inline canisters/src/vts/src/lib.rs]: BEncode and BDecode are not needed here, we do not send these types through TCP API.

File: canisters/src/vts/src/lib.rs [hunk 2]
```diff
@@ -110,6 +115,15 @@ pub struct Telemetry {
...
+pub struct AggregatedTelemetry {
+    daily_gas_usage: HashMap<String, u128>,
+    weekly_gas_usage: HashMap<String, u128>,
+    monthly_gas_usage: HashMap<String, u128>,
+}
+
```
Comment (lavrd) [inline canisters/src/vts/src/lib.rs:120]: Let's remove this newline :)

File: canisters/src/vts/src/lib.rs [hunk 3]
```diff
@@ -159,6 +173,65 @@ struct AgreementConditions {
...
+            VEHICLES.with(|vehicles| {
+                vehicles.borrow_mut().insert(vehicle_id, vehicle);
+            });
+        }
+    });
+}
+
+#[ic_cdk::query(guard = is_user)]
+fn get_aggregated_telemetry(vehicle: Principal) -> VTSResult<AggregatedTelemetry> {
```
Comment (lavrd) [inline canisters/src/vts/src/lib.rs]: Let's pass here interval. Like make a enum { Daily, Weekly, Monthly }.

**Review by lavrd — CHANGES_REQUESTED** [review]

File: canisters/src/vts/src/lib.rs [hunk 4]
```diff
@@ -110,6 +115,21 @@ pub struct Telemetry {
...
+    monthly_gas_usage: HashMap<String, u128>,
+}
+impl_storable!(AggregatedTelemetry);
+
+#[derive(CandidType, Deserialize, Debug)]
+enum AggregationInterval {
+    Daily,
+    Weekly,
+    Monthly,
+}
```
Comment (Robka99) [inline canisters/src/vts/src/lib.rs]: @lavrd enum interval like this?

**Review by lavrd — CHANGES_REQUESTED** [review]
Update after #26

**Review by lavrd — CHANGES_REQUESTED** [review]
So as we have a method to pre-fill data for the canister. Can we use this method in tests and write a test that telemetry aggregation is working properly? I think we can implement separate canister method to accumulate data and use it in test to not wait timer, for example. And in timer method we can use it to not write this logic twice.

**Comment by Robka99** [comment]
> So as we have a method to pre-fill data for the canister. Can we use this method in tests and write a test that telemetry aggregation is working properly? I think we can implement separate canister method to accumulate data and use it in test to not wait timer, for example. And in timer method we can use it to not write this logic twice.

maybe we can use a method like this for testing:
```
fn accumulate_telemetry_data_now() -> Result<(), Error> {
    accumulate_telemetry_data()
}
```
so we dont have to wait for the timer.

in the test class we can use: `fill_predefined_telemetry()` and `accumulate_telemetry_data_now()`

**Comment by lavrd** [comment]
> > So as we have a method to pre-fill data for the canister. Can we use this method in tests and write a test that telemetry aggregation is working properly? I think we can implement separate canister method to accumulate data and use it in test to not wait timer, for example. And in timer method we can use it to not write this logic twice.
> 
> 
> 
> maybe we can use a method like this for testing:
> 
> ```
> 
> fn accumulate_telemetry_data_now() -> Result<(), Error> {
> 
>     accumulate_telemetry_data()
> 
> }
> 
> ```
> 
> so we dont have to wait for the timer.
> 
> 
> 
> in the test class we can use: `fill_predefined_telemetry()` and `accumulate_telemetry_data_now()`
> 
> 

Yea, sure)

**Review by lavrd — APPROVED** [review]

**Pull request merged**